### ISSUE TITLE
Wrong method documentation returned in textDocument/hover.

### ISSUE THREAD

**Issue by shahamran** [clangd/clangd#829]
Hi. I'm loving clangd so far, keep up the good work!

I encountered a small problem. Assume you have some method declaration with documentation in a header file, and its definition sits in a source file. If the definition is preceded by a comment line or inline comment, this text is used as the documentation in the hover text, instead of the one in the declaration.

For example:
```cpp
// in foo.h
class Foo {
  /// some documentation
  static void foo();
};

// in foo.cpp
#include "foo.h"

/*static*/ void Foo::foo() { return; }
//                   ^ hovering here returns "static" instead of "some documentation".

```
I think it's a bug since it doesn't happen when both declaration and definition are in the same file.

**Logs**
```log
V[15:51:33.384] <<< {"id":30,"jsonrpc":"2.0","method":"textDocument/hover","params":{"position":{"character":21,"line":2},"textDocument":{"uri":"file:///homes/ransh/Desktop/foo.cpp"}}}

I[15:51:33.384] <-- textDocument/hover(30)
V[15:51:33.384] ASTWorker running Hover on version 69 of /homes/ransh/Desktop/foo.cpp
I[15:51:33.387] --> reply:textDocument/hover(30) 2 ms
V[15:51:33.387] >>> {"id":30,"jsonrpc":"2.0","result":{"contents":{"kind":"markdown","value":"### static-method `Foo::foo`  \n\n---\n→ `void`  \nstatic  \n\n---\n```cpp\n// In Foo\nprivate: void Foo::foo()\n```"},"range":{"end":{"character":24,"line":2},"start":{"character":21,"line":2}}}}
```

**System information**
Output of `clangd --version`:
```
clangd version 13.0.0 (https://github.com/llvm/llvm-project b1777b04dc4b1a9fee0e7effa7e177892ab32ef0)
Features: linux+grpc
```
Editor/LSP plugin: VSCode
Operating system: Ubuntu 18.04

**Comment by tom-anders** [comment]
[Here](https://searchfox.org/llvm/source/clang-tools-extra/clangd/Hover.cpp#319) Hover currently discard information from the index if it already has found some "documentation" from the AST. So that would probably be the right place to add a check to catch stuff like this. Is it enough to just hardcode a regex that catches `/* static */ here, or are there other common cases where this happens?

**Comment by shahamran** [comment]
@tom-anders
Thanks a lot for taking the time to address this!
Where I work, it's common to have detailed documentation in the header file, and implementation details in the source file. When I hover on a function call in a source file, I almost always intend to read the documentation, and not the implementation details.
So in my opinion, it's much better to just add the documentation from the index, rather than regex-ing for specific common strings.

Thank you.

**Comment by tom-anders** [comment]
Maybe the behavior could also be made configurable? I think we'd need to hear some other opinions here before implementing anything.

**Comment by sam-mccall** [comment]
> it's common to have detailed documentation in the header file, and implementation details in the source file

Agree. All else equal, if we see a forward decl and a definition, and both have comments, I think we should prefer the comment from the forward decl.

> [Here](https://searchfox.org/llvm/source/clang-tools-extra/clangd/Hover.cpp#319) Hover currently discard information from the index if it already has found some "documentation" from the AST

Right, and I think this is interacting badly with the fact that the AST contains only main-file comments, not preamble ones: https://searchfox.org/llvm/rev/506e93687140a1be054a825ae08f512d3ebbdbf8/clang-tools-extra/clangd/CodeCompletionStrings.cpp#84

I think skipping the index query in this case was just us being conservative when the docs-from-index codepath was added. I don't think it's necessary, we should query the index anyway and prefer the comment from there - we do have logic to try to ensure the index ends up with the right comment, and in principle the index has global knowledge to make this decision.

I *hope* the index is already preferring the docs from the headers over the impl, if not that's a bug we should fix.

> Is it enough to just hardcode a regex that catches `/* static */ here

Agree "/* static */" is noise that can be discarded. It's not the real fix to this bug (we should be preferring the header comment) but is probably useful anyway (we should not show "static" even if there's no other comment available).

> Maybe the behavior could also be made configurable?

I'm not sure exactly what you want to configure, and haven't seen anything in this bug that seems to warrant an option, just multiple good suggestions for heuristics.

Our bar for config options should be high: the tradeoff is maintaining two codepaths forever + config/docs, in exchange we get to improve behavior for probably <10% of users. (If we choose the right default, less than half should benefit from changing it, and it's very unlikely a quarter of them actually will...)

**Comment by tom-anders** [comment]
> I _hope_ the index is already preferring the docs from the headers over the impl, if not that's a bug we should fix.

Sounds good, I can do a quick check whether that's actually the case. 

> > Is it enough to just hardcode a regex that catches `/* static */ here
> Agree "/* static */" is noise that can be discarded. It's not the real fix to this bug (we should be preferring the header comment) but is probably useful anyway (we should not show "static" even if there's no other comment available).

Right, looks like it would make sense to extend `looksLikeDocComment()` with logic for this. 

> > Maybe the behavior could also be made configurable?
> 
> I'm not sure exactly what you want to configure, and haven't seen anything in this bug that seems to warrant an option, just multiple good suggestions for heuristics.
> 
> Our bar for config options should be high: the tradeoff is maintaining two codepaths forever + config/docs, in exchange we get to improve behavior for probably <10% of users. (If we choose the right default, less than half should benefit from changing it, and it's very unlikely a quarter of them actually will...)

Makes perfect sense!

**Comment by FrostKiwi** [comment]
Have exactly this case with SDL_Image.
The dense documentation is in the header with the declaration:
https://github.com/libsdl-org/SDL_image/blob/98ed433a6c27345ea4ac2c498ba50879d5f92778/include/SDL3_image/SDL_image.h#L115

But the Implementation with the definition has a tiny comment as well:
https://github.com/libsdl-org/SDL_image/blob/98ed433a6c27345ea4ac2c498ba50879d5f92778/src/IMG.c#L99

As a result, when clangd still loads it pulls from the header with good info:

<img width="1919" height="990" alt="Image" src="https://github.com/user-attachments/assets/a9acb569-3cbe-4936-9260-0b92d6d5b86a" />

But once it finds the Definition, the LSP Hover that clangd gives becomes low quality:

<img width="1919" height="991" alt="Image" src="https://github.com/user-attachments/assets/ed8eed2c-580d-4975-b187-040b3986badb" />
